mrmustard/lab_dev/circuit_components.py

```diff
@@ -181,6 +191,48 @@ def __getitem__(self, idx: Union[Mode, Sequence[Mode]]):
         ret._parameter_set = self.parameter_set
         return ret
 
+    def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
+        r"""
+        Contracts ``self`` and ``other``, without adding adjoints.
+        """
+        # initialized the ``Wires`` of the returned component
+        wires_ret = self.wires @ other.wires
+
+        # find the indices of the wires being contracted on the bra side
+        bra_modes = set(self.wires.bra.output.modes).intersection(other.wires.bra.input.modes)
+        idx_z = self.wires[bra_modes].bra.output.indices
+        idx_zconj = other.wires[bra_modes].bra.input.indices
+
+        # find the indices of the wires being contracted on the ket side
+        ket_modes = set(self.wires.ket.output.modes).intersection(other.wires.ket.input.modes)
+        idx_z += self.wires[ket_modes].ket.output.indices
+        idx_zconj += other.wires[ket_modes].ket.input.indices
+
+        # convert Bargmann -> Fock if needed
+        LEFT = self.representation
+        RIGHT = other.representation
+        if isinstance(LEFT, Bargmann) and isinstance(RIGHT, Fock):
+            raise ValueError("Cannot contract objects with different representations.")
+            # shape = [s if i in idx_z else None for i, s in enumerate(other.representation.shape)]
```

Comment:
where is the conversion taking place now?

Reply:
Right now we don't support conversions. If matmul finds that the representations are different, right now it will error. The error will be removed once we have the converters in place